controller/src/main/java/org/jboss/as/controller/operations/global/QueryOperationHandler.java

```diff
@@ -0,0 +1,253 @@
+            boolean isMatching = false;
+            List<Property> filterProperties = filter.asPropertyList();
+            List<Boolean> matches = new ArrayList<>(filterProperties.size());
+
+            for (Property property : filterProperties) {
+
+                final String filterName = property.getName();
+                final ModelNode filterValue = property.getValue();
+
+                boolean isEqual = false;
+
+                if(!filterValue.isDefined() || filterValue.asString().equals(UNDEFINED))  {
+                    // query for undefined attributes
+                    isEqual = !resource.get(filterName).isDefined();
+                }  else {
+
+                    final ModelType targetValueType = resource.get(filterName).getType();
+
+                    try {
+                        // query for attribute values (throws exception when types don't match)
+                        switch (targetValueType) {
+                            case BOOLEAN:
+                                isEqual = filterValue.asBoolean() == resource.get(filterName).asBoolean();
+                                break;
+                            case LONG:
+                                isEqual = filterValue.asLong() == resource.get(filterName).asLong();
+                                break;
+                            case INT:
+                                isEqual = filterValue.asInt() == resource.get(filterName).asInt();
+                                break;
+                            case DOUBLE:
+                                isEqual = filterValue.asDouble() == resource.get(filterName).asDouble();
+                                break;
+                            default:
+                                isEqual = filterValue.equals(resource.get(filterName));
+                        }
+                    } catch (IllegalArgumentException e) {
+                        throw new OperationFailedException("Illegal argument for attribute '"+filterName+"'. Expected type "+targetValueType);
+                    }
+
+                }
+
+                if(isEqual) {
+                    matches.add(resource.get(filterName).equals(filterValue));
+                }
+
+            }
+
+            if (Operator.AND.equals(operator)) {
+                // all matches must be true
+                isMatching = matches.size() == filterProperties.size();
+
+            } else if(Operator.OR.equals(operator)){
+                // at least one match must be true
+                for (Boolean match : matches) {
+                    if (match) {
+                        isMatching = true;
+                        break;
+                    }
+                }
+            }
+            else {
+                throw new OperationFailedException("Invalid operator"+operator);
+            }
```

Comment:
This would need i18n if it remained, but it can go. L99's OPERATOR_ATT.validateOperation(operation); has already validated the input, and you've already done Operator.valueOf.

If you want to leave this block here as a guard against adding new Operator values but forgetting to handle them, please throw IllegalStateException or some other RuntimeException. OperationFailedException indicates a user mistake, not a server fault. Places I do that I don't include any text in the exception message, as that would require i18n and that's overkill. The stack trace points to where the code is broken.